File: client/src/api/accounts.ts
import { apiClient } from './client'

export type AccountType = 'Checking' | 'Savings' | 'CreditCard'

export const accountTypes: AccountType[] = ['Checking', 'Savings', 'CreditCard']

export interface Account {
  id: number
  name: string
  type: AccountType
  isActive: boolean
}

export const accountsApi = {
  fetchAll: (includeInactive = false) =>
    apiClient.get<Account[]>(`/accounts${includeInactive ? '?includeInactive=true' : ''}`),
  create: (name: string, type: AccountType) => apiClient.post<Account>('/accounts', { name, type }),
  update: (id: number, name: string, type: AccountType) => apiClient.put<Account>(`/accounts/${id}`, { name, type }),
  deactivate: (id: number) => apiClient.post<void>(`/accounts/${id}/deactivate`),
}

File: src/Api/Features/Accounts/Accounts.cs
using SharpGrip.FluentValidation.AutoValidation.Endpoints.Extensions;

namespace Api.Features.Accounts;

public static partial class Accounts
{
    public enum AccountType
    {
        Checking,
        Savings,
        CreditCard,
    }

    public class Account
    {
        public int Id { get; set; }
        public required string Name { get; set; }
        public AccountType Type { get; set; }
        public bool IsActive { get; set; } = true;
        public required string CreatedByUserId { get; set; }
    }

    public static IServiceCollection AddAccountsFeature(this IServiceCollection services) => services
        .AddCreate()
        .AddUpdate()
        .AddDeactivate()
        .AddFetchAll()
        .AddFetchOne();

    public static IEndpointRouteBuilder MapAccountsFeature(this IEndpointRouteBuilder endpoints)
    {
        var group = endpoints.MapGroup("accounts")
            .WithTags("Accounts")
            .AddFluentValidationAutoValidation()
            .RequireAuthorization();
        group.MapCreate().MapUpdate().MapDeactivate().MapFetchAll().MapFetchOne();
        return endpoints;
    }
}
